### PR TITLE
[ntuple] Improve handling of large numbers of fields

### PR DESCRIPTION
**Pull request by jblomer**
Several fixes to meta-data handling that grew quadratic in the number of fields. The RNTuple limits tests is adjusted from 10k to 40k fields (30s --> 2s).

Finding pages in the RPagePool is a remaining contributor to a noticeable slowdown with a growing number of fields. That will be addressed in a follow-up PR.

### CODE REVIEW

**Review by pcanal — APPROVED** [review]
LGTM

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx [hunk 1]
```diff
@@ -844,6 +846,27 @@ ROOT::Experimental::Internal::RNTupleDescriptorBuilder::AddColumn(RColumnDescrip
    return RResult<void>::Success();
 }
 
+ROOT::Experimental::RResult<void>
+ROOT::Experimental::Internal::RNTupleDescriptorBuilder::AttachColumn(DescriptorId_t fieldId,
+                                                                     DescriptorId_t logicalColumnId)
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RNTupleDescriptor.cxx]: Why do we need a second function for this (that we can potentially forget to call), can we just do this at the end of `AddColumn` which should already have all necessary information?

(drive-by comment: can we de-duplicate the two `AddColumn` implementations? They seem to be largely doing the same, modulo error checking, in sometimes different order...)

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx [hunk 2]
```diff
@@ -291,11 +291,12 @@ ROOT::Experimental::RNTupleDescriptor::FindFieldId(std::string_view fieldName) c
 ROOT::Experimental::DescriptorId_t
 ROOT::Experimental::RNTupleDescriptor::FindLogicalColumnId(DescriptorId_t fieldId, std::uint32_t columnIndex) const
 {
-   for (const auto &cd : fColumnDescriptors) {
-      if (cd.second.GetFieldId() == fieldId && cd.second.GetIndex() == columnIndex)
-         return cd.second.GetLogicalId();
-   }
-   return kInvalidDescriptorId;
+   decltype(fFieldDescriptors)::const_iterator itr = fFieldDescriptors.find(fieldId);
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RNTupleDescriptor.cxx:290]: `auto`?

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx [hunk 3]
```diff
@@ -844,6 +846,27 @@ ROOT::Experimental::Internal::RNTupleDescriptorBuilder::AddColumn(RColumnDescrip
...
+   if (itrFieldDesc == fDescriptor.fFieldDescriptors.end()) {
+      return R__FAIL("AttachColumn: invalid field ID");
+   }
+   const auto itrColumnDesc = fDescriptor.fColumnDescriptors.find(logicalColumnId);
+   if (itrColumnDesc == fDescriptor.fColumnDescriptors.cend()) {
+      return R__FAIL("AttachColumn: invalid column ID");
+   }
+   const auto columnIdx = itrColumnDesc->second.GetIndex();
+   if (itrFieldDesc->second.fLogicalColumnIds.size() <= columnIdx) {
+      itrFieldDesc->second.fLogicalColumnIds.resize(columnIdx + 1);
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RNTupleDescriptor.cxx]: Do we need to initialize the intermediate elements with `kInvalidDescriptorId`?
```c++
itrFieldDesc->second.fLogicalColumnIds.resize(columnIdx + 1, kInvalidDescriptorId);
```

File: tree/ntuple/v7/inc/ROOT/REntry.hxx [hunk 4]
```diff
@@ -65,18 +65,38 @@ private:
...
    REntry() = default;
    explicit REntry(std::uint64_t modelId) : fModelId(modelId) {}
 
-   void AddValue(RFieldBase::RValue &&value) { fValues.emplace_back(std::move(value)); }
+   void UpdateValueIdxSortedByName(const std::string &fieldName)
+   {
+      auto itrSucc =
+         std::lower_bound(fValueIdxSortedByName.begin(), fValueIdxSortedByName.end(), fieldName,
+                          [this](size_t i, const std::string &f) { return CompValueIdxSortedByName(i, f); });
+      fValueIdxSortedByName.insert(itrSucc, fValues.size());
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/REntry.hxx:81]: Having a sorted `fValueIdxSortedByName` makes it possible to implement `GetToken` in logarithmic time, but the insertion sort is now quadratic at construction time, even if the user doesn't call `GetToken` (for example because they use the default `REntry`). A hashed `std::unordered_map` should get both construction and lookup down to linear complexity, is there a downside of going that way instead?

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx (hunk 1 above)
Comment (hahnjo) [reply tree/ntuple/v7/src/RNTupleDescriptor.cxx]: (ah, I see you do this in the third commit; please consider squashing its changes into the first commit)

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx (hunk 3 above)
Comment (hahnjo) [reply tree/ntuple/v7/src/RNTupleDescriptor.cxx]: (ah, I see you do this in the third commit; please consider squashing its changes into the first commit, we should not have potentially problematic code in the history)

**Review by hahnjo — CHANGES_REQUESTED** [review]
